velox/functions/prestosql/URLFunctions.h

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <boost/regex.hpp>
+#include <cctype>
+#include "folly/Uri.h"
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions {
+
+class UrlParser {
+ protected:
+  StringView submatch(const boost::cmatch& match, int idx);
+
+  template <typename TInString>
+  bool parse(TInString rawUrl, boost::cmatch& match);
+
+  bool authorityAndPathMatch(
+      boost::cmatch& urlMatch,
+      boost::cmatch& authAndPathMatch,
+      boost::cmatch& authorityMatch,
+      bool& hasAuthority);
+};
+
+FOLLY_ALWAYS_INLINE StringView
+UrlParser::submatch(const boost::cmatch& match, int idx) {
+  const auto& sub = match[idx];
+  return StringView(sub.first, sub.length());
+}
+
+template <typename TInString>
+bool UrlParser::parse(TInString rawUrl, boost::cmatch& match) {
+  static const boost::regex uriRegex(
+      "([a-zA-Z][a-zA-Z0-9+.-]*):" // scheme:
+      "([^?#]*)" // authority and path
+      "(?:\\?([^#]*))?" // ?query
+      "(?:#(.*))?"); // #fragment
+
+  if (UNLIKELY(!boost::regex_match(rawUrl.data(), match, uriRegex))) {
+    return false;
+  }
+
+  return true;
+}
+
+template <typename T>
+struct UrlExtractProtocolFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
```

Comment:
nit: it might be cleaner to use StringView for out_type<Varchar> and arg_type<Varchar>.